src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToAsyncMethods.cs

```diff
@@ -0,0 +1,244 @@
+                        // There are many kinds of blocks, so only break if we found a method symbol for this block.
+                        // Otherwise, blocks inside anonymous or local functions would not be detected - those would be the parent of the current operation.
+                        break;
+                    }
+                }
+
+                currentOperation = currentOperation.Parent;
+            }
+
+            return declaration != null;
+        }
+
+        // Check if the method only takes one ct and is the last parameter in the method signature.
+        // We want to compare the current method signature to any others with the exact same arguments in the exact same order.
+        private static bool VerifyMethodOnlyHasOneCancellationTokenAsLastArgument(
+            INamedTypeSymbol cancellationTokenType,
+            IMethodSymbol methodDeclaration,
+            out string? cancellationTokenParameterName)
+        {
+            cancellationTokenParameterName = null;
+
+            if (methodDeclaration.Parameters.Count(x => x.Type.Equals(cancellationTokenType)) == 1 &&
+                methodDeclaration.Parameters.Last() is IParameterSymbol lastParameter &&
+                lastParameter.Type.Equals(cancellationTokenType))
+            {
+                cancellationTokenParameterName = lastParameter.Name;
+            }
+
+            return cancellationTokenParameterName != null;
+        }
+
+        // Check if the currently used overload is the one that takes the ct, but is utilizing the default value offered in the method signature.
+        // We want to offer a diagnostic for this case, so the user explicitly passes the ancestor's ct.
+        private static bool InvocationIgnoresOptionalCancellationToken(INamedTypeSymbol cancellationTokenType, IInvocationOperation invocation, IMethodSymbol method)
+        {
+            if (method.Parameters.Any() && method.Parameters.Last() is IParameterSymbol lastParameter && lastParameter.Type.Equals(cancellationTokenType))
+            {
+                // If the ct parameter has a default value, return true if a value is not being explicitly passed in the invocation
+                return lastParameter.IsOptional && // Has a default value
+                       invocation.Arguments.Last() is IArgumentOperation lastArgument &&
+                       lastArgument.IsImplicit; // The default value is being used
+            }
+
+            return false;
+        }
+
+        // Check if there's a method overload with the same parameters as this one, in the same order, plus a ct at the end.
+        private static bool MethodHasCancellationTokenOverload(ITypeSymbol cancellationTokenType, IMethodSymbol method)
+        {
+            IMethodSymbol? overload = method.ContainingType.GetMembers(method.Name).OfType<IMethodSymbol>().FirstOrDefault(x => HasSameParametersPlusCancellationToken(cancellationTokenType, method, x));
```

Comment:
What you have probably has better performance than what we do in VSTHRD103, but FWIW we also catch opportunities in *extension methods* by using `LookupSymbols`:

https://github.com/microsoft/vs-threading/blob/fcf983e74a6d3f0000dbc03a62c0efc9783ce1aa/src/Microsoft.VisualStudio.Threading.Analyzers/VSTHRD103UseAsyncOptionAnalyzer.cs#L109-L113

But that's on the semantic model. I don't know if it's available without touching that.

Reply:
Agreed, `LookupSymbols` can be expensive here, so lets wait for customer feedback for important scenarios where this can happen.

Reply:
I'll add "extension methods" to the list of "future potential improvements".